### PR TITLE
Add profile toggle value to optionally show your e-mail address to authenticated users

### PR DESCRIPTION
**Pull request by joancodes**
# Notable Changes

* This PR aims to add the show_email boolean attribute to user settings for users to toggle if they want to publicise their email addresses to other authenticated users.

## Screenshots (Optional)

profile-with-email Screenshot

show-email-setting Screenshot

## Todo

* N/A

### CODE REVIEW

**Comment by pushcx** [comment]
Mentioning #1207 so GitHub links them

File: app/views/settings/index.html.erb [hunk 1]
```diff
@@ -181,6 +181,14 @@
 
     <h2>Miscellaneous Settings</h2>
```
Comment (pushcx) [inline app/views/settings/index.html.erb:183]: Please remove this `if`, all users should be able to check this box.

File: db/schema.rb [hunk 2]
```diff
@@ -222,7 +222,7 @@
     t.string "title", limit: 150, default: "", null: false
     t.text "description", size: :medium
     t.text "body", size: :medium
-    t.timestamp "created_at", default: -> { "current_timestamp() ON UPDATE current_timestamp()" }, null: false
+    t.timestamp "created_at", default: -> { "CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP" }, null: false
```
Comment (pushcx) [inline db/schema.rb:225]: Please remove this unrelated change.

File: db/schema.rb [hunk 3]
```diff
@@ -365,6 +366,6 @@
...
...` where ((`votes`.`user_id` = `read_ribbons`.`user_id`) and (`votes`.`comment_id` = `comments`.`id`))) AS `current_vote_reason` from (((`read_ribbons` join `comments` on((`comments`.`story_id` = `read_ribbons`.`story_id`))) join `stories` on((`stories`.`id` = `comments`.`story_id`))) left join `comments` `parent_comments` on((`parent_comments`.`id` = `comments`.`parent_comment_id`))) where ((`read_ribbons`.`is_following` = 1) and (`comments`.`user_id` <> `read_ribbons`.`user_id`) and (`comments`.`is_deleted` = 0) and (`comments`.`is_moderated` = 0) and ((`parent_comments`.`user_id` = `read_ribbons`.`user_id`) or ((`parent_comments`.`user_id` is null) and (`stories`.`user_id` = `read_ribbons`.`user_id`))) and (`stories`.`score` >= 0) and (`comments`.`score` >= 0) and ((`parent_comments`.`id` is null) or ((`parent_comments`.`score` >= 0) and (`parent_comments`.`is_moderated` = 0) and (`parent_comments`.`is_deleted` = 0))) and exists(select 1 from (`votes` `f` join `comments` `c` on((`f`.`comment_id` = `c`.`id`))) where ((`f`.`vote` < 0) and (`f`.`user_id` = `parent_comments`.`user_id`) and (`c`.`user_id` = `comments`.`user_id`) and (`f`.`story_id` = `comments`.`story_id`))) is false)
```
Comment (pushcx) [inline db/schema.rb:369]: Please remove this unrelated change.

**Comment by joancodes** [comment]
Hi @pushcx, I have made the necessary changes.

**Review by epilys — CHANGES_REQUESTED** [review]
Does this show the e-mail on authenticated users only? I can't find a check anywhere.

Also the option could probably use a label that clarifies this won't be completely public.

File: app/views/users/show.html.erb [hunk 4]
```diff
@@ -100,6 +100,14 @@
     <div style="clear: both;"></div>
   <% end %>
 
+  <% if @user.is_admin? && @user.show_email? %>
```
Comment (joancodes) [inline app/views/users/show.html.erb:103]: Hi @pushcx, could this be the check you are referring to?

Comment (joancodes) [reply app/views/users/show.html.erb:103]: Also, I am up for suggestions on what text to use on the label.

Comment (pushcx) [reply app/views/users/show.html.erb:103]: I was referring to the setting page. This `if` should be `||` so admins always see emails.

File: app/views/settings/index.html.erb [hunk 5]
```diff
@@ -181,6 +181,12 @@
...
 
+    <div class="boxline">
+      <%= f.label :show_email, "Show Email Address:",
+        :class => "required" %>
+      <%= f.check_box :show_email %>
```
Comment (pushcx) [inline app/views/settings/index.html.erb:187]: Please add a `<span class=“hint indent”>` like on Username above, with the text “Only shown to logged-in users”

**Comment by joancodes** [comment]
Hi @pushcx, I have added some more changes. LMK if I have missed anything.

**Comment by pushcx** [comment]
@joancodes Thanks for contributing this feature and being patient with changes. I’m really glad to merge this today.

**Comment by joancodes** [comment]
> 🎉 kudos @joancodes I just enabled it on lobste.rs and it works!

Glad to hear this! Thank you for letting me work on this and looking forward to making more meaningful contributions to lobste.rs 🎉

**Comment by banteg** [comment]
this pr has introduced a privacy vulnerability that leaks every user's email.

any logged-in user could enable “show email” on their own account, visit another user’s profile, and see that other user’s email even if the other user did not opt in.

originally reported here https://lobste.rs/s/7heurd/kyaaa_your_emails_are_showing_lobste_rs

fixed here https://github.com/lobsters/lobsters/commit/7037494eb85003603ecd87fb00bde994e111f103
